Profiler/config/Config.cpp

```diff
@@ -0,0 +1,112 @@
+#include "Config.h"
+#include "utils/WindowsUtils.h"
+#include <exception>
+
+std::string Config::getDefaultConfigPath()
+{
+	std::string profilerDllPath = WindowsUtils::getConfigValueFromEnvironment("PATH");
+	std::string profilerDllDirectory = StringUtils::removeLastPartOfPath(profilerDllPath);
+	return profilerDllDirectory + "\\Profiler.yml";
+}
+
+void Config::load(std::string configFilePath, std::string processPath, bool logProblemIfConfigFileDoesNotExist) {
+	this->processPath = processPath;
+	this->configPath = configFilePath;
+
+	if (WindowsUtils::isFile(configFilePath)) {
+		std::ifstream stream(configFilePath);
+		if (stream.fail()) {
+			problems.push_back("Failed to open the config file " + configFilePath + " for reading");
+			// we must still load the values from the environment in this case so we don't return here
+		}
+		else {
+			loadYamlConfig(stream);
+		}
+	}
+	else if (logProblemIfConfigFileDoesNotExist) {
+		problems.push_back("The config file " + configFilePath + " does not exist");
+	}
+
+	loadValues();
+}
+
+void Config::load(std::istream& configFileContents, std::string processPath) {
+	this->processPath = processPath;
+	loadYamlConfig(configFileContents);
+	loadValues();
+}
+
+void Config::loadYamlConfig(std::istream& configFileContents) {
+	try {
+		ConfigFile configFile = ConfigParser::parse(configFileContents);
+		apply(configFile);
+	}
+	catch (const std::exception& e) {
+		problems.push_back(std::string("Failed to parse the config file: ") + e.what());
+	}
+	catch (...) {
+		problems.push_back(std::string("Failed to parse the config file. The reason is unknown"));
+	}
+}
+
+void Config::loadValues()
+{
+	targetDir = getOption("targetdir");
+	enabled = getBooleanOption("enabled", true);
+	useLightMode = getBooleanOption("light_mode", false);
+	logAssemblyFileVersion = getBooleanOption("assembly_file_version", false);
+	logAssemblyPaths = getBooleanOption("assembly_paths", false);
+	dumpEnvironment = getBooleanOption("dump_environment", false);
+	ignoreExceptions = getBooleanOption("ignore_exceptions", false);
+	startUploadDaemon = getBooleanOption("upload_daemon", false);
+
+	std::string eagernessValue = getOption("eagerness");
+	if (eagernessValue.empty()) {
+		eagerness = 0;
+	}
+	else {
+		try {
+			eagerness = static_cast<size_t>(std::stoi(eagernessValue));
+		}
+		catch (...) {
+			problems.push_back("Invalid eagerness value configured: " + eagernessValue + ". Using the default of no eagerness instead");
+		}
+	}
+
+	disableProfilerIfProcessSuffixDoesntMatch();
+}
+
+void Config::disableProfilerIfProcessSuffixDoesntMatch() {
+	std::string processSuffix = WindowsUtils::getConfigValueFromEnvironment("process");
+	if (!processSuffix.empty() && !StringUtils::endsWithCaseInsensitive(processPath, processSuffix)) {
+		enabled = false;
+	}
+}
+
+void Config::apply(ConfigFile configFile) {
+	for (ProcessSection section : configFile.sections) {
+		if (std::regex_match(processPath, section.processRegex)) {
```

Comment:
suffix matching is inherently very limited and actually only useful for matching against the executable name in full.

I wanted to allow matching on the entire path of the executable in case I want to e.g. profile multiple applications that reside in the same directory. The other option I was thinking about was ANT patterns but that'd require us to reimplement them in C++ which is error-prone and takes a lot of effort. Thus regexes.

How about we rename this key to `executablePathRegex` and add an `executableName` key that does a case-insensitive equality check against the executable name (e.g. `foobar.exe`). That way the simple scenarios can be handled without requiring regexes and anything that's more complicated can still be done?

If you agree, please add a comment and I'll implement it this way

Reply:
Sounds good to me.